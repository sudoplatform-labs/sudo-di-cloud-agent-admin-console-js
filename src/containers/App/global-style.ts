import { createGlobalStyle } from 'styled-components';
import { LinkButton } from '../../components/LinkButton';
import { theme } from '../../theme';

export const GlobalStyle = createGlobalStyle`
  body {
    margin: 0;
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto',
      'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans',
      'Helvetica Neue', sans-serif;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
    background-color: ${theme.greys.athens};
  }

  a {
    color: ${theme.colors.sudoBlue};
    transition: color ease 0.25s;
    &:hover {
      text-decoration: underline;
    }
  }


  /**********
  * DropDown
  */
 
  .ant-dropdown-menu-item  {
    /* LinkButton should appear the same as <a> links in dropdown menus */
    ${LinkButton} {
      display: block;
      margin: -5px -12px;
      padding: 5px 12px;
      transition: all 0.3s;
      width: calc(100% + 24px);
      text-align: left;
    }
  }

  .ant-dropdown-trigger {
    line-height: normal
  }

  /**********
  * Button
  */

  .ant-btn-primary {
    background-color: ${theme.colors.sudoBlue};
    border-color: ${theme.colors.sudoBlue};
    color: ${theme.greys.alabastor};
  }

  .ant-btn-default {
    background-color: #fff;
    border-color: ${theme.colors.sudoBlue};
    color: ${theme.colors.sudoBlue};
  }

  .ant-btn-danger {
    background-color: ${theme.colors.coral};
  }

  /**********
  * Modal
  */

  .ant-modal {
    .ant-modal-header {
      border-bottom: none;
      padding: 24px 24px 8px;
    }

    .ant-modal-title {
      font-size: 16px;
    }

    .ant-modal-close {
      display: none;
    }
  }

  .ant-modal.ant-modal-confirm {
    .ant-modal-body {
      padding: 24px;

      .ant-modal-confirm-title {
        margin: -8px -24px 0;
        padding: 0 24px 18px;
      }

      .ant-modal-confirm-btns {
        float: none;
        display: flex;
        justify-content: flex-end;
        margin: 0 -24px -8px;
        padding: 18px 24px 0;
      }
    }
  }

  /**********
  * Form
  */

  .ant-form {
    text-align: left;

    legend {
      margin-bottom: 4px;
      border: none;
      color: ${theme.greys.woodsmoke};
      font-size: 14px;
    }
  }

  /**********
  * Descriptions
  */

  .ant-descriptions-item-label.ant-descriptions-item-colon {
    background-color: ${theme.greys.glamdring};
  }

  .ant-descriptions-bordered {
    .ant-descriptions-item-label {
      background-color: ${theme.greys.glamdring};
    }

    .ant-descriptions-item-label, .ant-descriptions-item-content {
      border-right: solid 1px ${theme.greys.geyser};
    }

    .ant-descriptions-item-label, tr:not(:last-child) .ant-descriptions-item-content {
      border-bottom: solid 1px ${theme.greys.geyser};
    }
  }

  .ant-descriptions-bordered .ant-descriptions-view {
    border: solid 1px ${theme.greys.geyser};
  }

  /**********
  * Table
  */

  .ant-table-tbody > tr .ant-table {
    margin: 0;
  }

  .ant-table thead tr th {
    background-color: ${theme.greys.glamdring};
  }

  tr.ant-table-expanded-row {
    background-color: ${theme.greys.glamdring};
  }

  .ant-table-thead tr th {
    box-shadow: inset 0 -1px 0 0 ${theme.greys.geyser};
    background-color: ${theme.greys.glamdring};
  }

  .ant-table-tbody > tr.ant-table-expanded-row > td {
    padding: 16px 24px;
    background-color: ${theme.greys.glamdring};
  }

  tr.ant-table-row {
    box-shadow: inset 0 -1px 0 0 #d4dbe3;
  }

  /**********
  * Tag
  */

  .ant-tag.ant-tag-success, .ant-tag.ant-tag-error, .ant-tag.ant-tag-default {
    border-radius: 4px;
    font-weight: 500;
  }
  .ant-tag.ant-tag-success {
    color: ${theme.colors.sudoBlue};
    border: solid 1px ${theme.colors.sudoBlue};
    background-color: rgba(0, 113, 222, 0.05); 
  }

  .ant-tag.ant-tag-error {
    color: ${theme.colors.coral};
    border: solid 1px ${theme.colors.coral};
    background-color: rgba(255, 67, 74, 0.05); 
  }

  .ant-tag.ant-tag-default {
    color: #8e95a7;
    border: solid 1px #9faab6;
    background-color: #ececf0; 
  }

  /**********
   * Tabs
   */

   .ant-tabs-tabpane {
     outline: none
   }
   
`;
