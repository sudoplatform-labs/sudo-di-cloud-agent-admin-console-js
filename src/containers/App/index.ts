export * from './App';
export * from './AppContext';
