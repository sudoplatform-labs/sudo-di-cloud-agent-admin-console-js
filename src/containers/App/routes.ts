import { createRouteRedirect } from '../../components/RouteRenderer';
import { RouteDefinitionFactory } from '../../routes';
import { Console } from '../Console';

export const getRoot: RouteDefinitionFactory = () => ({
  path: '/',
  component: createRouteRedirect('/console'),
  routes: [
    {
      path: '/console',
      component: Console,
      exact: false,
    },
  ],
});
