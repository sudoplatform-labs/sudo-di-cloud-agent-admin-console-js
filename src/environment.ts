export const uiConfig = '/uiconfig.json';
export const acapyConfig = '/acapy.json';

export interface EnvironmentInfo {
  acapyAdminUri: string; // REST API For Cloud Agent
  acapyAdminKey: string; // REST API authorisation key
  featureFlags: string[]; // Enabled/Disbled features for this customer
}
