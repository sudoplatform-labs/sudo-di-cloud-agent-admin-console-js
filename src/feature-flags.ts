/**
 * Feature flags are set during ACC deployment via runtime deployer variables.
 * E.g. RD_FF_EXAMPLE_FEATURE => FeatureFlag.EXAMPLE_FEATURE
 */
export enum FeatureFlag {
  EXAMPLE_FEATURE,
}

export function getFeatureFlags(featureFlagStrings: string[]): FeatureFlag[] {
  return featureFlagStrings
    .map((ffs) => {
      return FeatureFlag[ffs as keyof typeof FeatureFlag];
    })
    .filter((flag) => flag !== undefined)
    .sort();
}
