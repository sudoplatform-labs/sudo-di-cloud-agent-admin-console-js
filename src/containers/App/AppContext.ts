import React from 'react';
import { EnvironmentInfo } from '../../environment';
import { FeatureFlag } from '../../feature-flags';
import {
  WalletApi,
  LedgerApi,
  ConnectionApi,
  CredentialDefinitionApi,
  CredentialsApi,
  IntroductionApi,
  IssueCredentialApi,
  SchemaApi,
  PresentProofApi,
  TrustpingApi,
  RevocationApi,
} from '@sudoplatform-labs/sudo-di-cloud-agent';

// The Cloud Agent SDK exports each API as a class which
// we want to be accessible to the data model layer of the
// application.
export type CloudAgentAPI = {
  wallet: WalletApi;
  ledger: LedgerApi;
  introductions: IntroductionApi;
  connections: ConnectionApi;
  ping: TrustpingApi;
  defineSchemas: SchemaApi;
  defineCredentials: CredentialDefinitionApi;
  issueCredentials: IssueCredentialApi;
  revocations: RevocationApi;
  credentials: CredentialsApi;
  proofs: PresentProofApi;
  httpOptionOverrides: {
    // Different Swagger SDK generators insert different HTTP
    // headers so we override ones that cause issues
    httpPostOptionOverrides: {};
  };
};

export interface AppState {
  environment: EnvironmentInfo;
  featureFlags: FeatureFlag[];
  cloudAgentAPIs: CloudAgentAPI;
}

export const AppContext = React.createContext<AppState>(undefined as never);
