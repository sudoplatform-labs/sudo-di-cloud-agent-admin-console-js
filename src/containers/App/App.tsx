import 'antd/dist/antd.css';
import { LoadingOutlined } from '@ant-design/icons';
import React, { Suspense } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { useAsync } from 'react-use';
import styled from 'styled-components';
import { RouteRenderer } from '../../components/RouteRenderer';
import { SimplePage } from '../../components/SimplePage';
import { acapyConfig, EnvironmentInfo, uiConfig } from '../../environment';
import { getFeatureFlags } from '../../feature-flags';
import { AppContext, CloudAgentAPI } from './AppContext';
import { GlobalStyle } from './global-style';
import { getRoot } from './routes';
import {
  ConnectionApi,
  Configuration,
  CredentialsApi,
  CredentialDefinitionApi,
  IntroductionApi,
  IssueCredentialV10Api,
  IssueCredentialV20Api,
  LedgerApi,
  PresentProofV10Api,
  PresentProofV20Api,
  RevocationApi,
  SchemaApi,
  TrustpingApi,
  WalletApi,
} from '@sudoplatform-labs/sudo-di-cloud-agent';

const Container = styled.div`
  width: 100%;
  min-height: 100vh;
  min-width: 300px;
  display: flex;
  flex-direction: column;
`;

/**
 * The <App> container provides the minimum context needed
 * to launch the app, including the environment config
 * (obtained via `environment.json`); and the Decentralized
 * Identity Cloud Agent APIs needed to perform agent
 * functions such as credential issuance, schema definition
 * etc.
 *
 */
export const App: React.FC = () => {
  const initState = useAsync(async () => {
    // Get the environment information for this instance of
    // the app.  The environment files are either served from the
    // web server root or when running a local yarn dev environment
    // will be pulled from the top level "public" directory.
    const uiConfigResponse = await fetch(uiConfig);
    const ourConfig = await uiConfigResponse.json();

    const acapyConfigResponse = await fetch(acapyConfig);
    const agentConfig = await acapyConfigResponse.json();

    const environmentInfo: EnvironmentInfo = {
      acapyAdminUri: agentConfig.acapyAdminUri,
      acapyAdminKey: agentConfig.acapyAdminKey,
      featureFlags: ourConfig.featureFlags,
    };

    // Create all the Decentralized Identity Cloud Agent APIs
    // along with the API Key and base URL to talk to the
    // agent.
    const cloudAgentConfiguration = new Configuration({
      apiKey: environmentInfo.acapyAdminKey,
      basePath: environmentInfo.acapyAdminUri,
    });

    const cloudAgentAPIs: CloudAgentAPI = {
      wallet: new WalletApi(cloudAgentConfiguration),
      ledger: new LedgerApi(cloudAgentConfiguration),
      introductions: new IntroductionApi(cloudAgentConfiguration),
      connections: new ConnectionApi(cloudAgentConfiguration),
      ping: new TrustpingApi(cloudAgentConfiguration),
      defineSchemas: new SchemaApi(cloudAgentConfiguration),
      defineCredentials: new CredentialDefinitionApi(cloudAgentConfiguration),
      issueV10Credentials: new IssueCredentialV10Api(cloudAgentConfiguration),
      issueV20Credentials: new IssueCredentialV20Api(cloudAgentConfiguration),
      revocations: new RevocationApi(cloudAgentConfiguration),
      credentials: new CredentialsApi(cloudAgentConfiguration),
      presentV10Proofs: new PresentProofV10Api(cloudAgentConfiguration),
      presentV20Proofs: new PresentProofV20Api(cloudAgentConfiguration),
      httpOptionOverrides: {
        httpPostOptionOverrides: {
          headers: { 'Content-Type': 'application/json' },
        },
      },
    };

    return {
      environmentInfo,
      featureFlags: getFeatureFlags(environmentInfo.featureFlags),
      cloudAgentAPIs,
    };
  }, []);

  // Wait for the initialisation to complete before rendering our top
  // level page, or error out.
  if (initState.error) {
    return <SimplePage className="app__error">Error loading app.</SimplePage>;
  } else if (initState.loading || !initState.value) {
    return (
      <SimplePage className="app__loading">
        <LoadingOutlined />
      </SimplePage>
    );
  } else {
    return (
      // Make the application context available to all components below
      // via the useContext(AppContext) React hook.
      // This needs to include access to the
      // Cloud Agent APIs, this instances runtime environment and any
      // feature flags.
      <AppContext.Provider
        value={{
          environment: initState.value.environmentInfo,
          featureFlags: initState.value.featureFlags,
          cloudAgentAPIs: initState.value.cloudAgentAPIs,
        }}>
        <GlobalStyle />
        <BrowserRouter>
          <Container>
            <Suspense fallback={<LoadingOutlined />}>
              <RouteRenderer routeFactory={getRoot} />
            </Suspense>
          </Container>
        </BrowserRouter>
      </AppContext.Provider>
    );
  }
};
