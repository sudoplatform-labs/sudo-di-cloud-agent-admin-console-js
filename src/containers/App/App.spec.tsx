import { shallow } from 'enzyme';
import * as React from 'react';
import { mockComponent } from '../../test/mock-component';
import * as ReactUse from 'react-use';
import { uiConfig, EnvironmentInfo } from '../../environment';
import { acwMount } from '../../test/enzyme';
import { App } from './App';
import { AppContext } from './AppContext';
import * as FeatureFlags from '../../feature-flags';
import {
  WalletApi,
  LedgerApi,
  IntroductionApi,
  ConnectionApi,
  TrustpingApi,
  SchemaApi,
  CredentialDefinitionApi,
  IssueCredentialApi,
  RevocationApi,
  CredentialsApi,
  PresentProofApi,
} from '@sudoplatform-labs/sudo-di-cloud-agent';

const mockEnvironment: EnvironmentInfo = {
  acapyAdminUri: 'localhost:3000/',
  acapyAdminKey: 'testkey',
  featureFlags: ['EXAMPLE_FEATURE'],
};

const fetchSpy = jest.spyOn(window, 'fetch');
mockComponent('../../components/RouteRenderer/RouteRenderer', 'RouteRenderer');

AppContext['Provider'] = function MockAppContextProvider(props: any) {
  return <div>{props.children}</div>;
} as any;

describe('containers/App', () => {
  beforeEach(() => {
    fetchSpy.mockResolvedValue({
      json: jest.fn().mockResolvedValue(mockEnvironment),
    } as any);
  });

  it('should render', async () => {
    jest.spyOn(ReactUse, 'useAsync').mockReturnValueOnce({
      loading: false,
      error: undefined,
      value: {
        environmentInfo: 'ENVIRONMENT',
        featureFlags: [FeatureFlags.FeatureFlag.EXAMPLE_FEATURE],
      },
    });
    const wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should show loading state', async () => {
    fetchSpy.mockReturnValueOnce(new Promise(() => undefined));

    const wrapper = await acwMount(<App />);
    expect(wrapper.find('.app__loading').exists()).toBe(true);
  });

  it('should handle environment.json load error', async () => {
    fetchSpy.mockRejectedValue(new Error('ERROR!'));

    const wrapper = await acwMount(<App />);
    expect(wrapper.find('.app__error').exists()).toBe(true);
  });

  it('should set AppContext state', async () => {
    const wrapper = await acwMount(<App />);

    expect(fetchSpy).toBeCalledWith(uiConfig);
    expect(wrapper.find(AppContext.Provider).prop('value')).toMatchObject({
      environment: mockEnvironment,
      featureFlags: [FeatureFlags.FeatureFlag.EXAMPLE_FEATURE],
      cloudAgentAPIs: {
        wallet: expect.any(WalletApi),
        ledger: expect.any(LedgerApi),
        introductions: expect.any(IntroductionApi),
        connections: expect.any(ConnectionApi),
        ping: expect.any(TrustpingApi),
        defineSchemas: expect.any(SchemaApi),
        defineCredentials: expect.any(CredentialDefinitionApi),
        issueCredentials: expect.any(IssueCredentialApi),
        revocations: expect.any(RevocationApi),
        credentials: expect.any(CredentialsApi),
        proofs: expect.any(PresentProofApi),
      },
    });
  });
});
