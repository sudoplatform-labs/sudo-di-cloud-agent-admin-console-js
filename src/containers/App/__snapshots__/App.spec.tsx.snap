// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/App should render 1`] = `
<MockAppContextProvider
  value={
    Object {
      "cloudAgentAPIs": undefined,
      "environment": "ENVIRONMENT",
      "featureFlags": Array [
        0,
      ],
    }
  }
>
  <Memo(l) />
  <BrowserRouter>
    <styled.div>
      <Suspense
        fallback={<ForwardRef(LoadingOutlined) />}
      >
        <MockRouteRenderer
          routeFactory={[Function]}
        />
      </Suspense>
    </styled.div>
  </BrowserRouter>
</MockAppContextProvider>
`;
